yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,309 @@
+          const newGasPrice = await this.getGasPrice(gasConfig, attempts);
+
+          this.logger?.debug(
+            `L1 Transaction ${currentTxHash} appears stuck. Attempting speed-up ${attempts}/${gasConfig.maxAttempts} ` +
+              `with new priority fee ${formatGwei(newGasPrice.maxPriorityFeePerGas)} gwei`,
+          );
+
+          currentTxHash = await this.walletClient.sendTransaction({
+            ...request,
+            nonce,
+            gas: params.gasLimit,
+            maxFeePerGas: newGasPrice.maxFeePerGas,
+            maxPriorityFeePerGas: newGasPrice.maxPriorityFeePerGas,
+          });
+
+          txHashes.add(currentTxHash);
+          lastSeen = Date.now();
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      } catch (err: any) {
+        this.logger?.warn(`Error monitoring tx ${currentTxHash}:`, err);
+        if (err.message?.includes('reverted')) {
+          throw err;
+        }
+        await sleep(gasConfig.checkIntervalMs!);
+      }
+    }
+  }
+
+  /**
+   * Sends a transaction and monitors it until completion
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The receipt of the successful transaction
+   */
+  public async sendAndMonitorTransaction(
+    request: L1TxRequest,
+    gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<TransactionReceipt> {
+    const { txHash, gasLimit } = await this.sendTransaction(request, gasConfig);
+    return this.monitorTransaction(request, txHash, { gasLimit }, gasConfig);
+  }
+
+  /**
+   * Gets the current gas price with bounds checking
+   */
+  private async getGasPrice(_gasConfig?: L1TxUtilsConfig, attempt: number = 0): Promise<GasPrice> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const block = await this.publicClient.getBlock({ blockTag: 'latest' });
+    const baseFee = block.baseFeePerGas ?? 0n;
+
+    // Get initial priority fee from the network
+    let priorityFee = await this.publicClient.estimateMaxPriorityFeePerGas();
+    if (attempt > 0) {
+      // Bump priority fee by configured percentage for each attempt
+      priorityFee = (priorityFee * (100n + (gasConfig.priorityFeeBumpPercentage ?? 20n) * BigInt(attempt))) / 100n;
+    }
+
+    const maxFeePerGas = gasConfig.maxGwei! * WEI_CONST;
+    const maxPriorityFeePerGas = priorityFee;
+
+    this.logger?.debug(
+      `Gas price calculation (attempt ${attempt}): baseFee=${formatGwei(baseFee)}, ` +
+        `maxPriorityFee=${formatGwei(maxPriorityFeePerGas)}, maxFee=${formatGwei(maxFeePerGas)}`,
+    );
+
+    return { maxFeePerGas, maxPriorityFeePerGas };
+  }
+
+  /**
+   * Estimates gas and adds buffer
+   */
+  private async estimateGas(account: Account, request: L1TxRequest, _gasConfig?: L1TxUtilsConfig): Promise<bigint> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const initialEstimate = await this.publicClient.estimateGas({ account, ...request });
```

Comment:
Viem throws here if the tx would revert (or so I believe). It'd be good if we catch that error, extract the revert reason (if possible), and throw it back to the caller in an easily digestable format. There's a `prettyLogViemErrorMsg` we could use for that.

Reply:
it does and it's caught by the propose methods in l1-publisher. I've made #10066 to track this as there's a larger need for consistency in how we handle the error in this case so it's properly recorded in metrics